lib/puppet/provider/package/puppet_gem.rb

```diff
@@ -14,4 +14,14 @@
   else
     commands :gemcmd => "/opt/puppetlabs/puppet/bin/gem"
   end
+
+  def uninstall
+    begin
+      super
+    ensure
+      Puppet.debug("Invalidating rubygems cache after uninstalling gem '#{resource[:name]}'")
+      @rubygems ||= Puppet::Util::RubyGems::Source.new
+      @rubygems.clear_paths
```

Comment:
I think you'll want `Puppet::Util::Autoload.gem_source.clear_paths` There will always be a `gem_source` available, it's just when running under bundler the `clear_path` method is a noop.

Reply:
ok, thanks!  I'll give this a go, (probably on Monday).  The patch I proposed worked for me, but TBH I didn't have much confidence it was the 'correct' fix.

Reply:
@joshcooper I've tested the change you suggested, (to use https://github.com/puppetlabs/puppet/blob/e9107d98e911f8b956061ab0d5a116c559b8b9a3/lib/puppet/util/autoload.rb#L21) and it looks good. Anything else I need to do?

Reply:
👍